### PR TITLE
feat: add signing certificate to envelope

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Adds a PEM-encoded certificate to the DSSE envelope's signatures. Testing was done to ensure
1. We can add and retrieve the signing cert (the verifier will need the retrieval)
2. Rekor's intoto entry can validate this "slightly off-spec" envelope.

e2e test using this generator code and my verifier PR: https://github.com/asraa/slsa-on-github-test/actions/runs/2516202385
```
$ go run . -artifact-path ~/Downloads/binary-linux-amd64 -provenance ~/Downloads/binary-linux-amd64.intoto.jsonl -source github.com/asraa/slsa-on-github-test
Verified signature against tlog entry index 2694513 at URL: https://rekor.sigstore.dev/api/v1/log/entries/bf2882983fc8b0652902eab8500b540e1bb2530f91c004a7d6f0c204d703ded7
Signing certificate information:
 {
	"caller": "asraa/slsa-on-github-test",
	"commit": "e8c34fe013f732e9beb61a94b4667e1d30cde82e",
	"job_workflow_ref": "/asraa/slsa-github-generator/.github/workflows/builder_go_slsa3.yml@refs/heads/add-cert-to-envelope",
	"trigger": "workflow_dispatch",
	"issuer": "https://token.actions.githubusercontent.com"
}
FAILED: SLSA verification failed: untrusted reusable workflow: asraa/slsa-github-generator/.github/workflows/builder_go_slsa3.yml
exit status 2

```

### CODE REVIEW

File: signing/envelope/envelope.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	// Create an envelope.Envelope.
+	envWithCert := &Envelope{
+		PayloadType: env.PayloadType,
+		Payload:     env.Payload,
+		Signatures:  []Signature{},
+	}
+
+	for _, sig := range env.Signatures {
+		envWithCert.Signatures = append(envWithCert.Signatures,
+			Signature{Sig: sig.Sig, KeyID: sig.KeyID, Cert: string(cert)})
```
Comment (laurentsimon) [inline signing/envelope/envelope.go:51]: no base64 encoding needed for the cert?

**Review by laurentsimon — APPROVED** [review]
Thanks. Maybe add a test to verify it works as expected

File: signing/envelope/envelope.go (hunk 1 above)
Comment (ianlewis) [reply signing/envelope/envelope.go:51]: It's a PEM encoded cert as noted by the comment above. The cosign cli fulcio client having a field name of `Cert` with `[]byte` type is a bit unfortunate. Should probably also be `CertPEM`.
https://github.com/sigstore/cosign/blob/97b051686a0a4fd43d78dda34c62db696340c0d1/cmd/cosign/cli/fulcio/fulcio.go#L184

Encoded by Fulcio here:
https://github.com/sigstore/fulcio/blob/23ee9bcc89178dc13781ef8c6c6fb933b5e7e01b/pkg/api/client.go#L147

**Review by ianlewis — APPROVED** [review]

**Comment by ianlewis** [comment]
+1 for adding a testn

File: signing/envelope/envelope.go (hunk 1 above)
Comment (laurentsimon) [reply signing/envelope/envelope.go:51]: Thanks for the info. I should have guessed since it's part of a JSON string :)

**Comment by asraa** [comment]
Added envelope tests! See PR description.

**Pull request merged**